File: 2019/CSharp/Day01.cs
using System.Collections.Generic;
using System.Linq;
using AdventOfCode.Common.Extensions;
using Xunit;
using Xunit.Abstractions;

namespace AdventOfCode.CSharp.Year2019
{
    public class Day01 : TestBase
    {
        public Day01(ITestOutputHelper output) : base(output, 1) { }


        private (List<int>, int?) GetTestData(string name, int part)
        {
            var input = Input.ReadLines(DAY, name)
                .Select(l => l.ToInt32())
                .ToList();

            var expected = Input.ReadLines(DAY, $"{name}-answer{part}")
                ?.FirstOrDefault()
                ?.ToInt32();

            return (input, expected);
        }


        [Theory]
        [InlineData("example1")]
        [InlineData("example2")]
        [InlineData("example3")]
        [InlineData("example4")]
        [InlineData("input")]
        public void Part1(string inputName)
        {
            var (input, expected) = GetTestData(inputName, 1);

            var value = input.Sum(CalcFuel);

            output.WriteLine($"Answer: {value}");

            Assert.Equal(expected, value);
        }

        [Theory]
        [InlineData("example2")]
        [InlineData("example3")]
        [InlineData("input")]
        public void Part2(string inputName)
        {
            var (input, expected) = GetTestData(inputName, 2);

            var value = input.Sum(CalcTotalFuel);

            output.WriteLine($"Answer: {value}");

            Assert.Equal(expected, value);
        }




        /// <summary>
        /// Calculates the Fuel needed for the mass.
        /// </summary>
        /// <param name="mass"></param>
        /// <returns></returns>
        int CalcFuel(int mass)
        {
            return (mass / 3) - 2;
        }


        /// <summary>
        /// Calculates the fuel need to lift the mass, including the fuel.
        /// </summary>
        /// <param name="mass"></param>
        /// <returns></returns>
        int CalcTotalFuel(int mass)
        {
            var fuelMass = CalcFuel(mass);
            if (fuelMass <= 0L)
                return 0;
            return fuelMass + CalcTotalFuel(fuelMass);
        }

    }
}

File: 2019/FSharp/Day01.fs
namespace AdventOfCode.FSharp.Year2019

open System
open Xunit
open FSharp.Common



module ``Day 1: The Tyranny of the Rocket Equation`` =
    let day = 1



    //-------------------------------------------------------------------------



    type Puzzle () =

        member private this.CalcFuel mass : int =
            (mass / 3) - 2


        member private this.CalcFuelCorrectly mass : int =
            let fuelMass = this.CalcFuel(mass)
            match fuelMass with
            | _ when fuelMass > 0 -> fuelMass + this.CalcFuelCorrectly(fuelMass)
            | _ -> 0


        // What is the sum of the fuel requirements?
        member this.RunPart1 (input:int list) =
            let totalFuel = input |> List.sumBy this.CalcFuel
            totalFuel


        // What is the sum of the fuel requirements?
        member this.RunPart2 (input:int list) =
            let totalFuel = input |> List.sumBy this.CalcFuelCorrectly
            totalFuel



    //-------------------------------------------------------------------------



    let private GetPuzzleInput (part:int) (name:string) =
        let input = 
            InputHelper.LoadLines(day, name)
            |> Seq.map int
            |> Seq.toList
        
        let answer =
            InputHelper.LoadAnswer(day, $"%s{name}-answer%i{part}")
            |> InputHelper.AsInt

        input, answer



    [<Theory>]
    [<InlineData("example1")>]
    [<InlineData("example2")>]
    [<InlineData("example3")>]
    [<InlineData("example4")>]
    [<InlineData("input")>]
    let Part1 (name:string) =
        let input, expected = GetPuzzleInput 1 name

        let actual = (new Puzzle()).RunPart1 input

        match expected with
        | None -> Assert.Null actual
        | _ -> Assert.Equal (expected.Value, actual)


    [<Theory>]
    [<InlineData("example2")>]
    [<InlineData("example3")>]
    [<InlineData("input")>]
    let Part2 (name:string) =
        let input, expected = GetPuzzleInput 2 name

        let actual = (new Puzzle()).RunPart2 input

        match expected with
        | None -> Assert.Null actual
        | _ -> Assert.Equal (expected.Value, actual)
